File: src/test/java/com/gemini/krakenbot/repository/PortfolioStatsRepositoryImplTest.java
package com.gemini.krakenbot.repository;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.gemini.krakenbot.model.PortfolioStats;
import com.gemini.krakenbot.repository.impl.PortfolioStatsRepositoryImpl;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.*;

class PortfolioStatsRepositoryImplTest {

    private PortfolioStatsRepository repository;
    private static final String TEST_FILE = "portfolio-stats.json";

    @BeforeEach
    void setUp() {
        repository = new PortfolioStatsRepositoryImpl(new ObjectMapper());
    }

    @AfterEach
    void tearDown() {
        File file = new File(TEST_FILE);
        if (file.exists()) {
            //noinspection ResultOfMethodCallIgnored
            file.delete();
        }
    }

    @Test
    void load_ReturnsZeroWhenFileDoesNotExist() {
        // Ensure file is gone
        //noinspection ResultOfMethodCallIgnored
        new File(TEST_FILE).delete();

        PortfolioStats stats = repository.load();
        assertNotNull(stats);
        assertEquals(BigDecimal.ZERO, stats.getAllTimeHigh());
    }

    @Test
    void saveAndLoad_PersistsData() {
        PortfolioStats stats = new PortfolioStats(new BigDecimal("1000.50"));
        repository.save(stats);

        PortfolioStats loaded = repository.load();
        assertNotNull(loaded);
        assertEquals(0, new BigDecimal("1000.50").compareTo(loaded.getAllTimeHigh()));
    }

    @Test
    void load_HandlesIOException() throws Exception {
        File file = new File(TEST_FILE);
        java.nio.file.Files.writeString(file.toPath(), "{invalid json}");
        
        PortfolioStats stats = repository.load();
        assertNotNull(stats);
        assertEquals(BigDecimal.ZERO, stats.getAllTimeHigh());
    }

    @Test
    void save_HandlesIOException() throws Exception {
        ObjectMapper mockMapper = org.mockito.Mockito.mock(ObjectMapper.class);
        org.mockito.Mockito.doThrow(new java.io.IOException("simulated error"))
            .when(mockMapper).writeValue(org.mockito.ArgumentMatchers.any(File.class), org.mockito.ArgumentMatchers.any());

        PortfolioStatsRepositoryImpl errRepository = new PortfolioStatsRepositoryImpl(mockMapper);
        PortfolioStats stats = new PortfolioStats(BigDecimal.TEN);
        
        assertDoesNotThrow(() -> errRepository.save(stats));
    }
}

File: src/test/kotlin/com/gemini/krakenbot/repository/PortfolioStatsRepositoryImplTest.kt
package com.gemini.krakenbot.repository

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.gemini.krakenbot.model.PortfolioStats
import com.gemini.krakenbot.repository.impl.PortfolioStatsRepositoryImpl
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import java.io.File
import java.io.IOException
import java.math.BigDecimal
import java.nio.file.Files

@Suppress("unused")
class PortfolioStatsRepositoryImplTest : StringSpec() {

    private val testFileName = "test-portfolio-stats.json"
    private lateinit var repository: PortfolioStatsRepositoryImpl
    private lateinit var objectMapper: ObjectMapper

    init {
        beforeTest {
            val f = File(testFileName)
            if (f.exists()) f.delete()
            objectMapper = jacksonObjectMapper()
            repository = PortfolioStatsRepositoryImpl(objectMapper)
            val field =
                PortfolioStatsRepositoryImpl::class.java.getDeclaredField("filePath")
            field.isAccessible = true
            field.set(repository, testFileName)
        }

        afterTest {
            val f = File(testFileName)
            if (f.exists()) f.delete()
        }

        "load_NonExistentFile_ReturnsZeroStats" {
            val stats = repository.load()
            stats.shouldNotBeNull()
            stats.allTimeHigh shouldBe BigDecimal.ZERO
        }

        "load_Success" {
            val stats = PortfolioStats(BigDecimal("1000.50"))
            repository.save(stats)

            val loaded = repository.load()
            loaded.shouldNotBeNull()
            loaded.allTimeHigh shouldBe BigDecimal("1000.50")
        }

        "load_HandlesIOException" {
            val file = File(testFileName)
            Files.writeString(file.toPath(), "{invalid json}")

            val stats = repository.load()
            stats.shouldNotBeNull()
            stats.allTimeHigh shouldBe BigDecimal.ZERO
        }

        "save_HandlesIOException" {
            val mockMapper = mockk<ObjectMapper>(relaxed = true)
            every {
                mockMapper.writeValue(
                    any<File>(),
                    any<Any>()
                )
            } throws IOException("simulated error")

            val errRepository = PortfolioStatsRepositoryImpl(mockMapper)
            val stats = PortfolioStats(BigDecimal.TEN)

            shouldThrow<IOException> { errRepository.save(stats) }
        }
    }
}
